packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,201 @@
+
+const isSameAstNode = (actualNode: unknown, expectedNode: unknown): boolean => {
+  if (actualNode === expectedNode) {
+    return true;
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode == 'object' &&
+    typeof expectedNode == 'object'
+  ) {
+    if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
+      if (actualNode.length != expectedNode.length) {
+        return false;
+      }
+      return !actualNode.some(
+        (nodeEle, index) => !isSameAstNode(nodeEle, expectedNode[index]),
+      );
+    }
+    const actualNodeKeys = Object.keys(actualNode).filter(
+      key => !astIgnoreKeys.includes(key),
+    );
+    const expectedNodeKeys = Object.keys(expectedNode).filter(
+      key => !astIgnoreKeys.includes(key),
+    );
+    if (actualNodeKeys.length !== expectedNodeKeys.length) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !Object.prototype.hasOwnProperty.call(expectedNode, actualNodeKey),
+      )
+    ) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !isSameAstNode(
+            actualNode[actualNodeKey as keyof typeof actualNode],
+            expectedNode[actualNodeKey as keyof typeof expectedNode],
+          ),
+      )
+    ) {
+      return false;
+    }
+    return true;
+  }
+  return false;
+};
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-duplicate-type-constituents',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow duplicate constituents of union or intersection types',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    messages: {
+      duplicate:
+        '{{type}} type member {{duplicated}} is duplicated with {{previous}}.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreIntersections: {
+            type: 'boolean',
+          },
+          ignoreUnions: {
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreIntersections: false,
+      ignoreUnions: false,
+    },
+  ],
+  create(context, [{ ignoreIntersections, ignoreUnions }]) {
+    const sourceCode = context.getSourceCode();
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateConstituents: {
+        duplicated: TSESTree.TypeNode;
+        duplicatePrevious: TSESTree.TypeNode;
+      }[] = [];
+
+      node.types.reduce<
+        {
+          node: TSESTree.TypeNode;
+          type: Type;
+        }[]
+      >((uniqConstituents, constituentNode) => {
+        const type = checker.getTypeAtLocation(
+          parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+        );
```

Comment:
The point of the AST comparisons is so that we don't have to call `checker.getTypeAtLocation` if we don't need. But now this code is always checking that API even before it compares the ASTs! We're going to need to make these `type`s _lazily_ computed.

* The code should perform the `isSameAstNode` check _before_ it calls to `checker.getTypeAtLocation`
* If no nodes fail `isSameAstNode` (e.g. `type T = A | A`) then `checker.getTypeAtLocation` should never be called
* `checker.getTypeAtLocation` can be slow, and poking through https://github.com/microsoft/TypeScript/blob/f576398194bc1c16da7cbf4c6bd23f3bf0f41eea/src/compiler/checker.ts#L45148 I don't believe it _fully_ caches its results (@bradzacher has previously informed me on its quirks?). 
    * You might want to make a `Map<TSESTree.Node, Type>` to cache types?

Reply:
I have taken your suggestion and changed the implementation as follows
- First compare values with `isSameAstNode` and do early return if duplicates are found
- Only if no duplicates are found in `isSameAstNode`, call `checker.getTypeAtLocation` to compare type objects
- Type objects are cached as `Map<TSESTree.Node, Type>`